src/main/scala/chisel3/util/Counter.scala

```diff
@@ -19,27 +19,47 @@ import chisel3.internal.naming.chiselName  // can't use chisel3_ version because
   *     ...
   *   }
   * }}}
-  *
-  * @param n number of counts before the counter resets (or one more than the
-  * maximum output value of the counter), need not be a power of two
   */
 @chiselName
-class Counter(val n: Int) {
-  require(n >= 0, s"Counter value must be nonnegative, got: $n")
-  val value = if (n > 1) RegInit(0.U(log2Ceil(n).W)) else 0.U
+class Counter private (r: Range) {
+  require(r.start >= 0 && r.end >= 0, s"Counter range must be positive, got: $r")
+
+  private val delta = math.abs(r.step).U
+  private val width = math.max(log2Up(r.start), log2Up(r.end)) + 1
+
+  /** Creates a counter with the specified number of steps.
+    *
+    * @param n number of steps before the counter resets
+    */
+  def this(n: Int) { this(0 until n) }
 
-  /** Increment the counter
+  /** The current value of the counter. */
+  val value = if (r.length > 1) RegInit(r.start.U(width.W)) else r.start.U
+
+  /** The range of the counter values. */
+  def range: Range = r
+
+  /** Increments the counter by a step.
     *
     * @note The incremented value is registered and will be visible on the next clock cycle
-    * @return whether the counter will wrap to zero on the next cycle
+    * @return whether the counter will wrap on the next clock cycle
     */
   def inc(): Bool = {
-    if (n > 1) {
-      val wrap = value === (n-1).U
-      value := value + 1.U
-      if (!isPow2(n)) {
-        when (wrap) { value := 0.U }
+    if (r.length > 1) {
+      val wrap = value === r.last.U
+
+      when (wrap) {
+        value := r.start.U
```

Comment:
I think the special case can be applied by sticking an `if (!(isPow2(r.end) && isPow2(r.step)))` here.

Reply:
Thanks. I ended up using something very similar to that conditional, but had to also check that the range starts at zero.